File: csharp/CsFind/CsFindLib/FileTypes.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using FileTypesDictionary = System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<System.Collections.Generic.Dictionary<string, object>>>;

namespace CsFindLib;

public enum FileType
{
	Unknown,
	Archive,
	Audio,
	Binary,
	Code,
	Font,
	Image,
	Text,
	Video,
	Xml
};

public class FileTypes
{
	private const string Archive = "archive";
	private const string Audio = "audio";
	private const string Binary = "binary";
	private const string Code = "code";
	private const string Font = "font";
	private const string Image = "image";
	private const string Text = "text";
	private const string Video = "video";
	private const string Xml = "xml";

	private readonly string _fileTypesResource;
	private readonly IDictionary<string, ISet<string>> _fileTypeExtDictionary;
	private readonly IDictionary<string, ISet<string>> _fileTypeNameDictionary;

	public FileTypes()
	{
		_fileTypesResource = EmbeddedResource.GetResourceFileContents("CsFindLib.Resources.filetypes.json");
		_fileTypeExtDictionary = new Dictionary<string, ISet<string>>();
		_fileTypeNameDictionary = new Dictionary<string, ISet<string>>();
		PopulateFileTypesFromJson();
	}

	private void PopulateFileTypesFromJson()
	{
		var filetypesDict = JsonSerializer.Deserialize<FileTypesDictionary>(_fileTypesResource);
		if (filetypesDict!.ContainsKey("filetypes"))
		{
			var filetypeDicts = filetypesDict["filetypes"];
			foreach (var filetypeDict in filetypeDicts)
			{
				if (filetypeDict.TryGetValue("type", out var typeValue))
				{
					var name = ((JsonElement)typeValue).GetString();
					if (filetypeDict.TryGetValue("extensions", out var extensionsValue))
					{
						var extensions = ((JsonElement)extensionsValue).EnumerateArray()
							.Select(x => "." + x.GetString());
						var extensionSet = new HashSet<string>(extensions);
						_fileTypeExtDictionary[name!] = extensionSet;
					}
					if (filetypeDict.TryGetValue("names", out var namesValue))
					{
						var names = ((JsonElement)namesValue).EnumerateArray()
							.Select(x => "" + x.GetString());
						var nameSet = new HashSet<string>(names);
						_fileTypeNameDictionary[name!] = nameSet;
					}
				}
			}
		}
	}

	public static FileType FromName(string name)
	{
		return string.IsNullOrEmpty(name)
			? FileType.Unknown
			: name.ToLowerInvariant() switch
			{
				Archive => FileType.Archive,
				Audio => FileType.Audio,
				Binary => FileType.Binary,
				Code => FileType.Code,
				Font => FileType.Font,
				Image => FileType.Image,
				Text => FileType.Text,
				Video => FileType.Video,
				Xml => FileType.Xml,
				_ => FileType.Unknown
			};
	}

	public FileType GetFileType(FilePath filePath)
	{
		// more specific first
		if (IsCodeFile(filePath)) return FileType.Code;
		if (IsArchiveFile(filePath)) return FileType.Archive;
		if (IsAudioFile(filePath)) return FileType.Audio;
		if (IsFontFile(filePath)) return FileType.Font;
		if (IsImageFile(filePath)) return FileType.Image;
		if (IsVideoFile(filePath)) return FileType.Video;
		// more general last
		if (IsXmlFile(filePath)) return FileType.Xml;
		if (IsTextFile(filePath)) return FileType.Text;
		if (IsBinaryFile(filePath)) return FileType.Binary;
		return FileType.Unknown;
	}

	public bool IsArchiveFile(FilePath filePath)
	{
		return  _fileTypeNameDictionary[Archive].Contains(filePath.Name)
		        || _fileTypeExtDictionary[Archive].Contains(filePath.Extension.ToLowerInvariant());
	}

	public bool IsAudioFile(FilePath filePath)
	{
		return  _fileTypeNameDictionary[Audio].Contains(filePath.Name)
		        || _fileTypeExtDictionary[Audio].Contains(filePath.Extension.ToLowerInvariant());
	}

	public bool IsBinaryFile(FilePath filePath)
	{
		return  _fileTypeNameDictionary[Binary].Contains(filePath.Name)
		        || _fileTypeExtDictionary[Binary].Contains(filePath.Extension.ToLowerInvariant());
	}

	public bool IsCodeFile(FilePath filePath)
	{
		return _fileTypeNameDictionary[Code].Contains(filePath.Name)
		       || _fileTypeExtDictionary[Code].Contains(filePath.Extension.ToLowerInvariant());
	}

	public bool IsFontFile(FilePath filePath)
	{
		return _fileTypeNameDictionary[Font].Contains(filePath.Name)
		       || _fileTypeExtDictionary[Font].Contains(filePath.Extension.ToLowerInvariant());
	}

	public bool IsImageFile(FilePath filePath)
	{
		return _fileTypeNameDictionary[Image].Contains(filePath.Name)
		       || _fileTypeExtDictionary[Image].Contains(filePath.Extension.ToLowerInvariant());
	}

	public bool IsTextFile(FilePath filePath)
	{
		return  _fileTypeNameDictionary[Text].Contains(filePath.Name) ||
		        _fileTypeExtDictionary[Text].Contains(filePath.Extension.ToLowerInvariant()) ||
		        _fileTypeNameDictionary[Code].Contains(filePath.Name) ||
		        _fileTypeExtDictionary[Code].Contains(filePath.Extension.ToLowerInvariant()) ||
		        _fileTypeNameDictionary[Xml].Contains(filePath.Name) ||
		        _fileTypeExtDictionary[Xml].Contains(filePath.Extension.ToLowerInvariant());
	}

	public bool IsVideoFile(FilePath filePath)
	{
		return _fileTypeNameDictionary[Video].Contains(filePath.Name)
		       || _fileTypeExtDictionary[Video].Contains(filePath.Extension.ToLowerInvariant());
	}

	public bool IsUnknownFile(FilePath filePath)
	{
		return GetFileType(filePath) == FileType.Unknown;
	}

	public bool IsXmlFile(FilePath filePath)
	{
		return  _fileTypeNameDictionary[Xml].Contains(filePath.Name)
		        || _fileTypeExtDictionary[Xml].Contains(filePath.Extension.ToLowerInvariant());
	}
}

File: javascript/jsfind/src/filetypes.js
/*
 * filetypes.js
 *
 * identifies file types (archive, binary, text, unknown)
 */

const common = require('./common');
const config = require('./config');
const {FileType} = require('./filetype');
const {FileUtil} = require('./fileutil');

class FileTypes {
    constructor() {
        this.fileTypeExtMap = {};
        this.fileTypeNameMap = {};

        const json = FileUtil.getFileContentsSync(config.FILE_TYPES_JSON_PATH, 'utf-8');
        let obj = JSON.parse(json);
        if (Object.prototype.hasOwnProperty.call(obj, 'filetypes') && Array.isArray(obj.filetypes)) {
            obj.filetypes.forEach(ft => {
                let typename = ft.type;
                let extensions = ft.extensions;
                this.fileTypeExtMap[typename] = common.setFromArray(extensions);
                if (Object.prototype.hasOwnProperty.call(ft, 'names')) {
                    this.fileTypeNameMap[typename] = common.setFromArray(ft.names);
                } else {
                    this.fileTypeNameMap[typename] = [];
                }
            });
        } else throw new Error("Invalid filetypes file: " + config.FILE_TYPES_JSON_PATH);
        this.fileTypeExtMap.text = [].concat(this.fileTypeExtMap.text, this.fileTypeExtMap.code, this.fileTypeExtMap.xml);
        this.fileTypeNameMap.text = [].concat(this.fileTypeNameMap.text, this.fileTypeNameMap.code, this.fileTypeNameMap.xml);
    }

    getFileType(fileName) {
        // most specific first
        if (this.isCodeFile(fileName))
            return FileType.CODE;
        if (this.isArchiveFile(fileName))
            return FileType.ARCHIVE;
        if (this.isAudioFile(fileName))
            return FileType.AUDIO;
        if (this.isFontFile(fileName))
            return FileType.FONT;
        if (this.isImageFile(fileName))
            return FileType.IMAGE;
        if (this.isVideoFile(fileName))
            return FileType.VIDEO;
        // most general last
        if (this.isXmlFile(fileName))
            return FileType.XML;
        if (this.isTextFile(fileName))
            return FileType.TEXT;
        if (this.isBinaryFile(fileName))
            return FileType.BINARY;
        return FileType.UNKNOWN;
    }

    getFileTypeAsync(fileName, cb) {
        try {
            // most specific first
            if (this.isCodeFile(fileName))
                return cb(null, FileType.CODE);
            if (this.isArchiveFile(fileName))
                return cb(null, FileType.ARCHIVE);
            if (this.isAudioFile(fileName))
                return cb(null, FileType.AUDIO);
            if (this.isFontFile(fileName))
                return cb(null, FileType.FONT);
            if (this.isImageFile(fileName))
                return cb(null, FileType.IMAGE);
            if (this.isVideoFile(fileName))
                return cb(null, FileType.VIDEO);
            // most general last
            if (this.isXmlFile(fileName))
                return cb(null, FileType.XML);
            if (this.isTextFile(fileName))
                return cb(null, FileType.TEXT);
            if (this.isBinaryFile(fileName))
                return cb(null, FileType.BINARY);
        } catch (err) {
            return cb(err);
        }
        return cb(null, FileType.UNKNOWN);
    }

    isArchiveFile(fileName) {
        return this.fileTypeNameMap.archive.indexOf(fileName) > -1 ||
            this.fileTypeExtMap.archive.indexOf(FileUtil.getExtension(fileName)) > -1;
    }

    isAudioFile(fileName) {
        return this.fileTypeNameMap.audio.indexOf(fileName) > -1 ||
            this.fileTypeExtMap.audio.indexOf(FileUtil.getExtension(fileName)) > -1;
    }

    isBinaryFile(fileName) {
        return this.fileTypeNameMap.binary.indexOf(fileName) > -1 ||
            this.fileTypeExtMap.binary.indexOf(FileUtil.getExtension(fileName)) > -1;
    }

    isCodeFile(fileName) {
        return this.fileTypeNameMap.code.indexOf(fileName) > -1 ||
            this.fileTypeExtMap.code.indexOf(FileUtil.getExtension(fileName)) > -1;
    }

    isFontFile(fileName) {
        return this.fileTypeNameMap.font.indexOf(fileName) > -1 ||
            this.fileTypeExtMap.font.indexOf(FileUtil.getExtension(fileName)) > -1;
    }

    isImageFile(fileName) {
        return this.fileTypeNameMap.image.indexOf(fileName) > -1 ||
            this.fileTypeExtMap.image.indexOf(FileUtil.getExtension(fileName)) > -1;
    }

    isTextFile(fileName) {
        return this.fileTypeNameMap.text.indexOf(fileName) > -1 ||
            this.fileTypeExtMap.text.indexOf(FileUtil.getExtension(fileName)) > -1;
    }

    isVideoFile(fileName) {
        return this.fileTypeNameMap.video.indexOf(fileName) > -1 ||
            this.fileTypeExtMap.video.indexOf(FileUtil.getExtension(fileName)) > -1;
    }

    isXmlFile(fileName) {
        return this.fileTypeNameMap.xml.indexOf(fileName) > -1 ||
            this.fileTypeExtMap.xml.indexOf(FileUtil.getExtension(fileName)) > -1;
    }

    isUnknownFile(fileName) {
        return this.getFileType(fileName) === FileType.UNKNOWN;
    }

    static fromName(name) {
        switch (name.toLowerCase()) {
            case 'archive':
                return FileType.ARCHIVE;
            case 'audio':
                return FileType.AUDIO;
            case 'binary':
                return FileType.BINARY;
            case 'code':
                return FileType.CODE;
            case 'font':
                return FileType.FONT;
            case 'image':
                return FileType.IMAGE;
            case 'text':
                return FileType.TEXT;
            case 'video':
                return FileType.VIDEO;
            case 'xml':
                return FileType.XML;
            default:
                return FileType.UNKNOWN;
        }
    }

    static toName(fileType) {
        switch (fileType) {
            case FileType.ARCHIVE:
                return 'archive';
            case FileType.AUDIO:
                return 'audio';
            case FileType.BINARY:
                return 'binary';
            case FileType.CODE:
                return 'code';
            case FileType.FONT:
                return 'font';
            case FileType.IMAGE:
                return 'image';
            case FileType.TEXT:
                return 'text';
            case FileType.VIDEO:
                return 'video';
            case FileType.XML:
                return 'xml';
            default:
                return 'unknown';
        }
    }

    static fileTypesToString(name, fileTypes) {
        if (fileTypes.length) {
            let s = `${name}=[`;
            for (let i=0; i < fileTypes.length; i++) {
                if (i > 0) s += ', ';
                s += FileTypes.toName(fileTypes[i]);
            }
            s += ']';
            return s;
        }
        return `${name}=[]`;
    }
}

exports.FileTypes = FileTypes;
